### ISSUE TITLE
`mg_get_server_ports` has poor code

### ISSUE THREAD

**Issue by mulle-nat** [civetweb/civetweb#845]
[mg_get_server_ports](https://github.com/civetweb/civetweb/blob/master/src/civetweb.c#L3638)

```
int
mg_get_server_ports(const struct mg_context *ctx,
                    int size,
                    struct mg_server_port *ports)
{
	int i, cnt = 0;

	if (size <= 0) {
		return -1;
	}
```
First up, I don't see  an easy way to figure out what to pass in as `size`. The API should IMO return `ctx->num_listening_sockets` if  `size` is 0.  That is sort of conventional.  The check for negative `size` would be superflous, if `size` were `unsigned int` (or `size_t` as in the old API). The return value as `int` is OK.

So the proposed change would be:

```
int
mg_get_server_ports(const struct mg_context *ctx,
                    unsigned int size,
                    struct mg_server_port *ports)
{
	int i, cnt = 0;

	if (size == 0) {
		return( ctx->num_listening_sockets);
	}
```

The [mg_server_port API](https://github.com/civetweb/civetweb/blob/master/docs/api/mg_server_port.md) says, that there are three possible value for `ports->protocol`  namely "IPv4" is **1**, "IPv6" is **2** and "IPv4 and IPv6" is **3**`. But the function returns **1** and **3** only. 

```

		if (ctx->listening_sockets[i].lsa.sa.sa_family == AF_INET) {
			/* IPv4 */
			ports[cnt].protocol = 1;
			cnt++;
		} else if (ctx->listening_sockets[i].lsa.sa.sa_family == AF_INET6) {
			/* IPv6 */
			ports[cnt].protocol = 3;
			cnt++;
		}
```

The return value **3** indicates "IPv4 and IPv6", which is clearly not the case (it should be "IPv6" therefore **2**).

`cnt` is only incremented in the case if `AF_INET` and `AF_INET6` actually match. If its sure that either must match, the second `if` is superflous and confusing.  And this code would be better:

```
		if (ctx->listening_sockets[i].lsa.sa.sa_family == AF_INET) {
			/* IPv4 */
			ports[cnt].protocol = 1;
		} 
                else {
			/* IPv6 */
			ports[cnt].protocol = 2; 
		}
                cnt++;
```

If it is possible though, that **_both_** `if`s fail, then the `ports[cnt]` will get clobbered by the next `ports[cnt]` due to the missing `cnt++`. 

I guess the author really wanted to write this:

```
		if (ctx->listening_sockets[i].lsa.sa.sa_family == AF_INET) {
			/* IPv4 */
			ports[cnt].protocol += 1;
		} 
                if (ctx->listening_sockets[i].lsa.sa.sa_family == AF_INET6) {
			/* IPv6 */
			ports[cnt].protocol += 2;
		}
                cnt++;
```

**Comment by bel2125** [comment]
> First up, I don't see an easy way to figure out what to pass in as size. The API should IMO return ctx->num_listening_sockets if size is 0. That is sort of conventional. 

If you are using a lot of ports, that might indeed be beneficial. For a typical small number (~2 ports), you might just use some size, e.g. 10 elements.

The typical use case for this function is to specify "0" as port number, let the operating system choose, and then ask for the port using this function. In this case you know the number of ports, so it seems noone missed this feature up to now.

But I agree, we can add this.

> The check for negative size would be superflous, if size were unsigned int (or size_t as in the old API). The return value as int is OK.

I want to get rid of "size_t" in the interfaces, since it causes a lot of problems when compiling for 32 and 64 bits and intermixing this with other data types. For data sizes (e.g., file sizes), "int64_t" is much more appropriate, since you can have a 5 GB file also on a 32 bit system.

Unsigned could be done, then you enter an unsigned size and return a signed size - I tend to stick with signed here.

> The mg_server_port API says, that there are three possible value for ports->protocol namely "IPv4" is 1, "IPv6" is 2 and "IPv4 and IPv6" is 3`. But the function returns 1 and 3 only.
> ...
> I guess the author really wanted to write this:
> ...

When looking at the code, I see yet another problem. From the description ("IPv4" is 1, "IPv6" is 2 and "IPv4 and IPv6" is 3`) I would expect something like "if there is a IPv4 socket with port 1234 and an IPv6 socket with port 1234, return one entry with protocol 3 for port 1234" but it doesn't work like this - and makes not much sense either.

```
		if (ctx->listening_sockets[i].lsa.sa.sa_family == AF_INET) {
			/* IPv4 */
			ports[cnt].protocol += 1;
		} 
                if (ctx->listening_sockets[i].lsa.sa.sa_family == AF_INET6) {
			/* IPv6 */
			ports[cnt].protocol += 2;
		}
                cnt++;
```

Since `sa_family` can be only either AF_INET or AF_INET6 and AF_INET != AF_INET6, the += is no better than the =

**Comment by bel2125** [comment]
Probably:

```C
int
mg_get_server_ports(const struct mg_context *ctx,
                    int size,
                    struct mg_server_port *ports)
{
	int i, cnt = 0;

	if ((size < 0) || ((size > 0) && ((ports == 0) || (ctx == 0)))) {
		return -1; /* invalid parameter */
	}

        if (size == 0) {
            return (int)(ctx->listening_sockets);
        }

	memset(ports, 0, sizeof(*ports) * (size_t)size);

	for (i = 0; (i < size) && (i < (int)ctx->num_listening_sockets); i++) {

		ports[cnt].port =
#if defined(USE_IPV6)
		    (ctx->listening_sockets[i].lsa.sa.sa_family == AF_INET6)
		        ? ntohs(ctx->listening_sockets[i].lsa.sin6.sin6_port)
		        :
#endif
		        ntohs(ctx->listening_sockets[i].lsa.sin.sin_port);
		ports[cnt].is_ssl = ctx->listening_sockets[i].is_ssl;
		ports[cnt].is_redirect = ctx->listening_sockets[i].ssl_redir;

switch (ctx->listening_sockets[i].lsa.sa.sa_family) {
    case AF_INET:
			/* IPv4 */
			ports[cnt].protocol = 1; 
    break;
    case AF_INET6:
			/* IPv6 */
			ports[cnt].protocol = 2;
    break;
   default:
      TODO: There are feature requests for other protocols, e.g. domain sockets
      In this case, set protocol to some other value
  }
			cnt++;
	}
	return cnt;
}


```

**Comment by mulle-nat** [comment]
I would prefer the size == 0 check before the invalid parameter, so it doesn't matter what's in `ports`, which could then e NULL.  Also there is no real use for the `cnt` variable. It's always the same value as `i`. 

Simplified:

```
   if( ctx->num_listening_sockets < size)
      size = ctx->num_listening_sockets;

	for (i = 0; i < size; i++) {
```
and then replace `cnt` with `i`.

**Comment by bel2125** [comment]
i is the loop variable - you should assume it's invalid outside the loop
cnt is the count of elements and the return value

In case there is a socket that is neither AF_INET nor AF_INET6, they differ and the proposed change would no longer work. Currently there are only AF_INET and AF_INET6, and there is no filtering in this function. But there it is likely to get some extensions here.

**Comment by bel2125** [comment]
> "IPv4" is 1, "IPv6" is 2 and "IPv4 and IPv6" is 3`

There is a difference between 2 (only IPv6) and 3 (IPv4+IPv6): The IPV6_V6ONLY socket option.
This is what the code should have checked.

**Comment by mulle-nat** [comment]
That's not how C works, if you declare it on top. :)

bel2125 schrieb:

**Comment by bel2125** [comment]
Yes it would be allowed in C, but it works differently in other languages.
Maybe that's why some consider it bad style to rely on this.
Also, I think it's a maintenance problem to use loop variables outside loops.

Anyway, do you think there is a real bug, or just "matter of taste" code style issues?

**Comment by mulle-nat** [comment]
Just a matter of taste :)
